working/terminating-tokens/feature-specification.md

```diff
@@ -0,0 +1,1665 @@
+the user wants an empty body (typically a loop where the condition expression
+has a side effect), they can use an empty block, which is usually more readable
+anyway.
+
+Without empty statements, we know that a newline inside a control flow statement
+(after the if and while condition, after `else`, etc.) *cannot* be meaningful.
+Thus, we can safely ignore it, which means that this:
+
+```dart
+if (condition)
+  body();
+```
+
+Continues to be parsed like:
+
+```dart
+if (condition) body();
+```
+
+Just like it is today. That means the 6,102 existing cases like this are not
+broken, at the expense of having to fix the 17 uses of empty statements.
+
+**TODO: The prototype doesn't currently implement this restriction, but there
+are only a handful of cases where it occurs.**
+
+#### Prefix and infix operators
+
+Dart has a couple of operators that can appear both at the beginning and middle
+of an expression. That leads to ambiguity if a newline appears before that
+operator, as in:
+
+```dart
+a
+-b
+```
+
+That could be parsed as either:
+
+```dart
+a;
+-b;
+```
+
+or:
+
+```dart
+a - b;
+```
+
+
+
+In all cases, we follow the eagerness principle and treat a newline as
+significant before these operators. That means one of these operators at the
+beginning of a line is treated like a unary operator if possible.
+
+Note that these are *only* a problem at the "top level" of a statement
+expression. Once the operator is nested inside parentheses or an argument list
+or something, a semicolon is never allowed and we know we can ignore the
+newline.
+
+The set of operators affected are:
+
+*   `-`, which can be used for both unary negation and infix subtraction:
+
+    ```dart
+    -negate
+    sub - tract
+    ```
+
+    There are **10,551** infix `-` operators in the corpus and **13** (0.123%)
+    are at the beginning of a line. Those thirteen are all in an expression
+    context where we know the newline cannot be significant, so are OK too.
+
+*   `<`, which can be used for generic collection literals and comparison:
+
+    ```dart
+    <int>[]
+    less < than
+    ```
+
+    There are **26,049** infix `<` operators and none at the beginning of a
+    line.
+
+*   `(`, which can be a parenthesized expression or a function invocation:
+
+    ```dart
+    (parenthesized)
+    function(call)
+    ```
+
+    There are **1,437,185** invocations in the corpus, none of which have the
+    `(` on the next line.
+
+*   `[`, which can be a list literal or an index operator:
+
+    ```dart
+    [list, literal]
+    list[access]
+    ```
+
+    There are **173,429** index operator calls in the corpus, **2** (0.001%) of
+    which have the `[` on the next line. Both are the result of chained index
+    operators like:
+
+    ```dart
+    exifData[tag] = ExifConstants.stringValues["Components"]
+            [exifData[tag][0]]
+    ```
+
+    One is in an expression context where the newline is ignored and one is not.
+
```

Comment:
There is also `--` and `++` which can both end and start an expression (prefix and suffix operators):
```dart
a
++
b
```
can be parsed in two ways with optional semicolons. 
I doubt that ever happens in practice, so using the eagerness principle should be safe.